### ISSUE TITLE
Support a streaming table

### ISSUE THREAD

**Issue by coretl** [PandABlocks/PandABlocks-server#61]
We have a requirement for 50kHz table based position compare. The SEQ block has 4096 rows, and though it is double buffered this would be 11 table writes a second, which gives very little margin for error. Long table support has much bigger buffers, but appending while reading is not currently supported. After discussing the possibilities with @EmilioPeJu and @Araneidae, we think:
- The FPGA side of the long table support (reading from a sequence of DMA blocks) seems suitable for this application, but may be lacking some flow control to say when it is ready for the next block
- There is a need to run PGEN in a stream mode too, but need to retain the current static mode
- If we can make long table support both stream and static mode, then maybe we don't need a short table mode

A possible long term plan is:
- Remove short table support and convert SEQ to use long table mode
- Remove table append support
- Table support (formerly long table support) gains a new `MODE` attribute that can be `fixed` or `fifo`
- In `fixed` mode it does what it currently does
- In `fifo` mode then the value is always empty, clearing of the fifo is done from the FPGA side, and writing to the table will append to the fifo

A plan of how to get there:
- Test how fast PGEN can consume the table
- Add the `MODE` attribute to table support on the table, and make writes append to it if in `fifo` mode
- Switch the mode, and run the PGEN test again but make sure the buffers are filled in time
- Then add the flow control and clearing to the server and PGEN
- Convert SEQ to use long tables
- Remove short table and append support from server

**Comment by coretl** [comment]
@tomtrafford

**Comment by EmilioPeJu** [comment]
Hi @coretl ,

This change involves both, the FPGA and the server repos, regarding the FPGA side here is the structure of my initial prototype:

![Image](https://github.com/user-attachments/assets/83c9bfca-3e29-4ecf-93ec-ee4e38e17943)

- SEQ manages the main state machine, a new state called RESETTING was added to allow the DMA pipeline to be flushed when a stream is suddenly disabled.
- SEQ RING TABLE is a fifo for the entries.
- SEQ DMA manages the DMA transfers and push the data to SEQ RING TABLE.

The decision to use interrupts (when the FPGA can receive a new buffer) was made for robustness.
The end of the streaming is signaled by sending a 0-length table (no address needed), this simplifies the implementation.

There are 2 main error conditions:
- Pushing a buffer when FPGA is not ready.
- SEQ state machine consuming more entries than the DMA can provide.
Both conditions will move the sequencer to IDLE updating HEALTH accordingly.

I have finished the initial FPGA implementation which was tested only in testbenches, now I'm working on the server side to be able to do tests on real hardware (and to confirm a minimum bandwidth of 16MB/s for each instance works).
Here is my initial workflow proposal for the server side:
- User sets STREAMING_MODE to 1 (Enabled).
- User writes a table (and possibly more tables) and set enable high.
- Every time the user writes a table, a DMA buffer is allocated and queued, there will be one queue per instance of block using the DMA interface.
- Every time a new buffer can be pushed, there will be an interrupt popping a queued buffer from the appropriate queue and pushing it to the FPGA, possibly de-allocating the buffer that the FPGA finished using.
- The user can check how many entries(or data words) are queued by reading a metadata field, TABLE_LEFT maybe? (I'm open to suggestions for a good field name), this information comes from an ioctl syscall to the driver.
- When the user wants to finish the stream, a table with 0 entries will be written.

The error cases (and disabling case) needs to be handled carefully, a possibility is interrupting again to clean up the driver's leftovers.

**Comment by coretl** [comment]
Sounds reasonable to me. Can I just check that STREAMING_MODE and TABLE_LEFT are attributes of a long table field? If so this seems fine to me. An alternative name to TABLE_LEFT would be BUFFERED_ROWS.

**Comment by EmilioPeJu** [comment]
BUFFERED_ROWS would be a long table metadata field. STREAMING_MODE is a register in the seq block, maybe we should rename it to TABLE_MODE and make it part of the table interface?

**Comment by coretl** [comment]
> BUFFERED_ROWS would be a long table metadata field. STREAMING_MODE is a register in the seq block, maybe we should rename it to TABLE_MODE and make it part of the table interface?

Yes, I think this would be best. In which case how about a `.MODE` enum attribute with `static` or `streaming` entries? @tomtrafford thoughts?

**Comment by EmilioPeJu** [comment]
OK, I changed the design again, I wanted to get rid of ambiguous situations and decided to remove registers `STREAMING_MODE` and `CAN_WRITE_NEXT`.
Table length now contains a flag to indicate if there will be more tables, if not, it will start using the repeat counts until the end.
There are 2 interrupts, one to indicate that next buffer can be pushed and one to indicate that operation ended (and remaining buffers can be cleaned up).
In the end, I made TABLE_LENGTH=0 a reset (which is what pgen does), because I found it convenient for the driver, so it can be sure of what state the FPGA is.
Regarding user interface, I was thinking `>>` append operation contains the more flag and `>` insert operation doesn't.
I'm now encapsulation all the table registers logic and dma transfer logic into one entity so that it can be reused in every block having the `dma` interface, I will leave the fifo out of it so the block can be more flexible.
Regarding testbenches, I've created a few ones that are useful for development, I think it might be useful to keep them somewhere (like a new folder under tests/dev or maybe the sim folder?) to help future developments.

**Comment by EmilioPeJu** [comment]
In both cases but you shouldn't have a falling edge of enable:
my proposal is that `<<` will signal the more flag,  then once you send a last table with `<`,  you shouldn't push more tables, either the sequencer needs to finish or you disable (falling edge of enable will reset everything).

**Comment by EmilioPeJu** [comment]
PRs were merged!
